synapse/storage/databases/main/stream.py

```diff
@@ -743,14 +743,17 @@ async def get_room_event_before_stream_ordering(
         """
 
         def _f(txn: LoggingTransaction) -> Optional[Tuple[int, int, str]]:
-            sql = (
-                "SELECT stream_ordering, topological_ordering, event_id"
-                " FROM events"
-                " WHERE room_id = ? AND stream_ordering <= ?"
-                " AND NOT outlier"
-                " ORDER BY stream_ordering DESC"
-                " LIMIT 1"
-            )
+            sql = """
+                SELECT stream_ordering, topological_ordering, event_id
+                FROM events
+                LEFT JOIN rejections USING (event_id)
+                WHERE room_id = ?
+                    AND stream_ordering <= ?
+                    AND NOT outlier
+                    AND rejections.reason IS NULL
```

Comment:
I'm having trouble understanding how #12319 broke things. The previous code did a `get_recent_events_for_room`, which looks like it doesn't check for rejection either?

Reply:
Oh, hmm, you're right. The PR did make it worse though as the query actually doubles the specified limit (to take account of filtering), so before we'd have to have two rejected events next to each other to hit the bug